src/Compilers/CSharp/Portable/Lowering/Instrumentation/DebugInfoInjector.cs

```diff
@@ -0,0 +1,371 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This type is responsible for adding debugging sequence points for the executable code.
+    /// It can be combined with other <see cref="Instrumenter"/>s. Usually, this class should be 
+    /// the root of the chain in order to ensure sound debugging experience for the instrumented code.
+    /// </summary>
+    internal partial class DebugInfoInjector : CompoundInstrumenter
+    {
+        /// <summary>
+        /// A singleton object that performs only one type of instrumentation - addition of debugging sequence points. 
+        /// </summary>
+        public static readonly DebugInfoInjector Singleton = new DebugInfoInjector(Instrumenter.NoOp);
+
+        public DebugInfoInjector(Instrumenter previous)
```

Comment:
No, it is my intent to keep the constructor accessible. For example, for cases when we need to combine sequence points with code coverage instrumentation.